index.ts

```diff
@@ -0,0 +1,476 @@
+	ignoreOffensive?: boolean;
+	ignorePositiveDefensive?: boolean;
+	ignorePositiveEvasion?: boolean;
+	ignoresBurn?: boolean;
+	ignoresDefenseBoosts?: boolean;
+	infiltrates?: boolean;
+	isFutureMove?: boolean;
+	isSpread?: boolean | 'allAdjacent';
+	isZ?: boolean | ID;
+	multiaccuracy?: boolean;
+	multihit?: number | number[];
+	multihitType?: 'parentalbond';
+	noCopy?: boolean;
+	noDamageVariance?: boolean;
+	noFaint?: boolean;
+	noMetronome?: string[];
+	noPPBoosts?: boolean;
+	noSketch?: boolean;
+	nonGhostTarget?: string;
+	ohko?: true | 'Ice' | null;
+	percentHealed?: number;
+	pressureTarget?: string;
+	priority?: number;
+	pseudoWeather?: string:
+	recoil?: [number, number];
+	secondaries?: SecondaryEffect[];
+	self?: SelfEffect;
+	selfBoost?: {boosts?: Partial<BoostsTable>};
+	selfSwitch?: ID | boolean;
+	selfdestruct?: string | boolean;
+	sideCondition?: string;
+	sleepUsable?: boolean;
+	slotCondition?: string;
+	spreadHit: boolean;
+	spreadModifier?: number;
+	stallingMove?: boolean;
+	status?: StatusName;
+	stealsBoosts?: boolean;
+	terrain?: string;
+	thawsTarget?: boolean;
+	useSourceDefensive?: boolean;
+	useTargetOffensive?: boolean;
+	usesHighestAttackStat?: boolean;
+	volatileStatus?: ID;
+	weather?: ID;
+	willCrit?: boolean;
+	zMoveBoosts?: Partial<BoostsTable>;
+	zMoveEffect?: string;
+	zMovePower?: number;
+}
+
+class Moves {
+	static forGen(gen: Generation): Promise<Moves>;
+
+	get(name: string): Move | undefined;
+	getByID(id: ID): Move | undefined;
+	next(): Move | true;
+}
+
+// species.ts
+
+type GenderName = 'M' | 'F' | 'N';
+type UsageTierName = 'OU' | 'UU' | 'RU' | 'NU' | 'PU';
+type TierName = UsageTierName | 'Uber' | 'UUBL' | 'RUBL' | 'NUBL' | 'PUBL';
+type EvoType = 'trade' | 'stone' | 'levelMove' | 'levelExtra' | 'levelFriendship' | 'levelHold';
+interface SpeciesAbilities = {0: string, 1?: string, H?: string, S?: string};
+type MoveSource = string;
+
+interface EventInfo {
+	generation: Generation;
+	level?: number;
+	shiny?: boolean | 1;
+	gender?: GenderName;
+	nature?: NatureName;
+	ivs?: Partial<StatsTable>;
+	perfectIVs?: number;
+	isHidden?: boolean;
+	abilities?: ID[];
+	maxEggMoves?: number;
+	moves?: ID[];
+	pokeball?: ID;
+	from?: string;
+}
+
+interface Species extends Data {
+	type 'Species';
+
+	abilities: SpeciesAbilities;
+	addedType?: string;
+	baseForme?: string;
+	baseSpecies?: string;
+	baseStats: StatsTable;
+	battleOnly?: boolean
+	canHatch?: boolean;
+	color: string;
+	cosmeticForms?: ID[];
+	doublesTier?: TierName;
+	eggGroups: string[];
+	encounters?: EventInfo[]
+	eventOnly?: boolean
+	eventPokemon?: EventInfo[];
+	evoCondition?: string;
+	evoItem?: string;
+	evoLevel?: number;
+	evoMove?: string;
+	evos?: ID[];
+	evoType?:
+	formeid: string;
+	formeLetter?: string;
+	forme?: string;
+	gender?: GenderName;
+	genderRatio?: {M: number, F: number};
+	heightm: number;
+	isMega?: boolean;
+	isPrimal?: boolean;
+	isTotem: boolean;
+	learnset?: {[moveid: string]: MoveSource[]};
```

Comment:
- potentially move to a diff file as we discussed in pm
- is `moveid` meant to be `ID` ?
- i'd prefer `Map`

Reply:
Yes, will split out alone with event data. `moveid` cant be an `ID` because `ID` cant be used to key an object, but they can key a `Map` which is all the more reason to use that datatype instead.

Reply:
One concern regarding `Map` - the client is specifically avoiding it currently: https://github.com/Zarel/Pokemon-Showdown-Client/blob/master/CONTRIBUTING.md#es2018, but I would just have it polyfill or drop support for browsers no one uses...

Reply:
89% of users have browsers that support `Map`, and 96% of desktop users, I don't see why we'd avoid a polyfill

Reply:
because Zarle is the Don Quixote of doing unnecessary work supporting people on IE7

Reply:
89% is far from 100%. We have around 100 staff members, so that means 11 of them would no longer be able to use PS (well, probably nowhere near that many; staff members are probably generally higher-tech than the userbase, but just to give an idea of the numbers involved).

Reply:
My main concern with `Map` is that it makes serialization/deserialization a lot harder, which might be bad considering the use-case.